File: java-pine/src/main/java/com/tipsy/pine/operators/ReorderSort.java
package com.tipsy.pine.operators;

import com.tipsy.pine.AbstractOperator;
import com.tipsy.pine.OperatorInput;
import com.tipsy.pine.OperatorOutput;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class ReorderSort extends AbstractOperator {
    private boolean ascending;

    @Override
    public void init(Map<String, Object> params) throws Exception {
        String order = (String) params.getOrDefault("order", "desc");
        switch (order) {
            case "asc":
                ascending = true;
                break;
            case "desc":
                ascending = false;
                break;
            default:
                throw new IllegalArgumentException("reorder_sort: unsupported order: " + order);
        }
    }

    @Override
    public void execute(OperatorInput input, OperatorOutput output) throws Exception {
        int n = input.itemCount();
        if (n == 0) return;

        String field = itemInput.get(0);
        List<int[]> entries = new ArrayList<>(n);
        double[] vals = new double[n];
        for (int i = 0; i < n; i++) {
            vals[i] = toDouble(input.item(i, field));
            entries.add(new int[]{i});
        }

        Integer[] indices = new Integer[n];
        for (int i = 0; i < n; i++) indices[i] = i;

        final double[] sortVals = vals;
        java.util.Arrays.sort(indices, (a, b) -> {
            int cmp = Double.compare(sortVals[a], sortVals[b]);
            return ascending ? cmp : -cmp;
        });

        List<Integer> order = new ArrayList<>(n);
        for (int idx : indices) {
            order.add(idx);
        }
        output.setItemOrder(order);
    }

    private static double toDouble(Object v) throws Exception {
        if (v instanceof Number) {
            return ((Number) v).doubleValue();
        }
        throw new Exception("cannot convert " + (v == null ? "null" : v.getClass().getName()) + " to double");
    }
}

File: pine-cpp/operators/reorder/reorder_sort.cpp
#include "pine/operator.hpp"

#include <algorithm>

#include "operators/_helpers.hpp"

namespace pine {

class ReorderSortOp : public Operator, public ConsumesRowSet, public MutatesRowSet {
 public:
  void init(const OperatorConfig& cfg) override {
    op_name_ = cfg.name;
    if (cfg.metadata.item_input.empty()) {
      throw ExecutionError("reorder_sort requires item_input field");
    }
    field_ = cfg.metadata.item_input.front();
    const auto& obj = cfg.params.as_object();
    auto it = obj.find("order");
    order_ = (it == obj.end()) ? std::string("desc") : it->second.as_string();
  }
  void execute(const OperatorInput& input, OperatorOutput& out) override {
    if (input.item_count() == 0) {
      return;
    }
    struct Keyed {
      double v;
      std::size_t idx;
    };
    std::vector<Keyed> keyed;
    keyed.reserve(input.item_count());
    // Batched column access: one lock + one lookup instead of per-element
    // item() calls.
    std::vector<Variant> raw = input.item_column(field_);
    for (std::size_t i = 0; i < raw.size(); ++i) {
      try {
        const Variant& v = raw[i];
        if (v.is_null()) {
          throw operators::OperatorError("required field \"" + field_ + "\" is nil on item[" +
                                         std::to_string(i) + "]");
        }
        keyed.push_back({operators::to_double(v), i});
      } catch (const operators::OperatorError& err) {
        throw ExecutionError("reorder_sort: item[" + std::to_string(i) + "]." + field_ + ": " + err.what());
      }
    }
    if (order_ == "asc") {
      std::stable_sort(keyed.begin(), keyed.end(), [](const Keyed& a, const Keyed& b) { return a.v < b.v; });
    } else if (order_ == "desc") {
      std::stable_sort(keyed.begin(), keyed.end(), [](const Keyed& a, const Keyed& b) { return a.v > b.v; });
    } else {
      throw ExecutionError("reorder_sort: unsupported order \"" + order_ + "\"");
    }
    std::vector<int> order_vec;
    order_vec.reserve(keyed.size());
    for (const auto& k : keyed) {
      order_vec.push_back(static_cast<int>(k.idx));
    }
    out.set_item_order(std::move(order_vec));
  }

 private:
  std::string op_name_;
  std::string field_;
  std::string order_;
};

static const OperatorSchema k_reorder_sort_schema{
    .name = "reorder_sort",
    .type = OpType::Reorder,
    .description = "Sorts items by a numeric field in ascending or descending order.",
    .params =
        {
            {"order",
             {.type = "string",
              .required = false,
              .default_value = Variant("desc"),
              .description = "Sort direction \xe2\x80\x94 \"asc\" or \"desc\"."}},
        },
    .metadata = {.common_input = "[]", .common_output = "[]", .item_input = "[<field>]", .item_output = "[]"},
};
PINE_REGISTER_OPERATOR_T(ReorderSortOp, k_reorder_sort_schema)

}  // namespace pine
